micropip/_commands/freeze.py

```diff
@@ -1,12 +1,110 @@
 import importlib.metadata
 import json
+import sys
 from copy import deepcopy
-from typing import Any
+from importlib.metadata import Distribution
+from typing import TypedDict
 
 from packaging.utils import canonicalize_name
+from packaging.version import Version
 
 from .._compat import REPODATA_INFO, REPODATA_PACKAGES
 from .._utils import fix_package_dependencies
+from ..package_index import query_package
+
+IN_VENV = sys.prefix != sys.base_prefix
+
+
+class PkgEntry(TypedDict):
+    name: str
+    version: str
+    file_name: str
+    install_dir: str
+    sha256: str | None
+    imports: list[str]
+    depends: list[str]
+
+
+def get_pkg_entry_micropip(dist: Distribution, url: str) -> PkgEntry:
+    name = dist.name
+    version = dist.version
+    sha256 = dist.read_text("PYODIDE_SHA256")
+    assert sha256
+    imports = (dist.read_text("top_level.txt") or "").split()
+    requires = dist.read_text("PYODIDE_REQUIRES")
+    if not requires:
+        fix_package_dependencies(name)
+        requires = dist.read_text("PYODIDE_REQUIRES")
+    if requires:
+        depends = json.loads(requires)
+    else:
+        depends = []
+
+    return dict(
+        name=name,
+        version=version,
+        file_name=url,
+        install_dir="site",
+        sha256=sha256,
+        imports=imports,
+        depends=depends,
+    )
+
+
+async def get_pkg_entry_pip(dist: Distribution) -> PkgEntry | None:
+    resp = await query_package(dist.name)
+    ver = resp.releases.get(Version(dist.version), None)
+    if ver is None:
+        return None
+    wheel = next(ver)
+    await wheel.download({})
+    requires = [req.name for req in wheel.requires(set())]
+    return dict(
+        name=dist.name,
+        version=dist.version,
+        file_name=wheel.url,
+        install_dir="site",
+        sha256=wheel.sha256,
```

Comment:
So IIUC, we are fetching the installed package again to get the URL and the sha256? I am not sure if is it a reliable behavior. What happens if a user installed packages from non-PyPI repositories or from URL?

Reply:
It works if they installed from url (at least in principle it can be made to work) because we can get the url from `direct_url.json`. The multiple indexes problem is bigger because we don't have metadata indicated which index we got it from. I think micropip should store the index url into maybe `INDEX`.

@pradyunsg @henryiii is there something I'm missing? If someone has a pip.conf with an extra-index-url or passes this as a cli instruction, is it possible to reconstruct which packages in the venv came from which indices? If not maybe we could make a packaging PEP adding INDEX to package metadata.